From f1e88262ef77bcd9d0ec50180ecd1a9c0f646049 Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Tue, 25 Jun 2019 10:34:14 +0800
Subject: [PATCH] dnfpayload.py: customize for OE

Upstream-Status: Inappropriate [oe specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>

Update context for anaconda 32.

the storage and data are not propagated to the payload again, so remove
from setup().

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 34 on 20210727.
* rebase code
* remove debug code
* remove patch for function language_groups() which has been remove from upstream

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 37 on 20230227.

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 38 on 20231107.

Signed-off-by: Kai Kang <kai.kang@windriver.com>


Rebase for anaconda 39 on 20240327.

* funtion is_environment_valid() has been removed in upstream
* create class OEConfigureTask to handle oe specific configuration
* set DNF default source from install disk

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Create link file /lib before install packages since command `depmod`
creates it as a directory rather than a link file.

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 42 on 20250328.

Signed-off-by: Kai Kang <kai.kang@windriver.com>
---
 pyanaconda/modules/payloads/payload/dnf/dnf.py     |   5 +
 .../modules/payloads/payload/dnf/dnf_manager.py    | 185 +++++++++++++++++----
 .../modules/payloads/payload/dnf/initialization.py |   4 +
 .../modules/payloads/payload/dnf/installation.py   |  74 ++++++++-
 .../modules/payloads/payload/dnf/requirements.py   |  21 +++
 .../payloads/payload/dnf/transaction_progress.py   |   9 +-
 pyanaconda/modules/payloads/payload/dnf/utils.py   |  10 --
 .../modules/payloads/payload/dnf/validation.py     |  83 +++++++++
 pyanaconda/payload/dnf/payload.py                  |   4 +-
 9 files changed, 351 insertions(+), 44 deletions(-)

diff --git a/pyanaconda/modules/payloads/payload/dnf/dnf.py b/pyanaconda/modules/payloads/payload/dnf/dnf.py
index e37fedd64c..e6f6c7abaa 100644
--- a/pyanaconda/modules/payloads/payload/dnf/dnf.py
+++ b/pyanaconda/modules/payloads/payload/dnf/dnf.py
@@ -51,6 +51,7 @@ from pyanaconda.modules.payloads.payload.dnf.installation import (
     SetRPMMacrosTask,
     UpdateDNFConfigurationTask,
     WriteRepositoriesTask,
+    OEConfigureTask,
 )
 from pyanaconda.modules.payloads.payload.dnf.tear_down import ResetDNFManagerTask
 from pyanaconda.modules.payloads.payload.dnf.utils import (
@@ -496,6 +497,10 @@ class DNFModule(PayloadBase):
                 configuration=self.packages_configuration,
                 dnf_manager=self.dnf_manager,
             ),
+            OEConfigureTask(
+                sysroot=conf.target.system_root,
+                dnf_module=self,
+            ),
             ResetDNFManagerTask(
                 dnf_manager=self.dnf_manager
             )
diff --git a/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py b/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py
index a759196012..ab448f845c 100644
--- a/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py
+++ b/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py
@@ -21,6 +21,10 @@ import multiprocessing
 import shutil
 import threading
 import traceback
+import os
+import inspect
+import collections
+import configparser
 
 import dnf
 import dnf.exceptions
@@ -42,7 +46,7 @@ from pyanaconda.core.constants import (
 )
 from pyanaconda.core.i18n import _
 from pyanaconda.core.payload import ProxyString, ProxyStringError
-from pyanaconda.core.util import get_os_release_value
+from pyanaconda.core.util import get_os_release_value, execWithRedirect, execWithCapture
 from pyanaconda.modules.common.errors.installation import PayloadInstallationError
 from pyanaconda.modules.common.errors.payload import (
     UnknownCompsEnvironmentError,
@@ -120,11 +124,17 @@ class DNFManager:
         self._md_hashes = {}
         self._enabled_system_repositories = []
 
+        # OE specific
+        self.image = {}
+        self.tasks = {}
+        self.failed_scriptlets_pkgnames = collections.OrderedDict()
+
     @property
     def _base(self):
         """The DNF base."""
         if self.__base is None:
             self.__base = self._create_base()
+            self._oe_configure(self.__base.conf)
 
         return self.__base
 
@@ -153,7 +163,7 @@ class DNFManager:
 
         # Set the platform id based on the /os/release present
         # in the installation environment.
-        platform_id = get_os_release_value("PLATFORM_ID")
+        platform_id = get_os_release_value("VERSION_ID")
 
         if platform_id is not None:
             base.conf.module_platform_id = platform_id
@@ -187,6 +197,9 @@ class DNFManager:
         self._md_hashes = {}
         self._enabled_system_repositories = []
 
+        self.image = {}
+        self.tasks = {}
+
         log.debug("The DNF base has been reset.")
 
     def configure_base(self, data: PackagesConfigurationData):
@@ -241,7 +254,8 @@ class DNFManager:
 
         :return: a list of ids
         """
-        return [env.id for env in self._base.comps.environments]
+        log.info("%s %s: %s" % (self.__class__.__name__, inspect.stack()[0][3], self.tasks.keys()))
+        return sorted(self.tasks.keys())
 
     def _get_environment(self, environment_name):
         """Translate the given environment name to a DNF object.
@@ -252,7 +266,7 @@ class DNFManager:
         if not environment_name:
             return None
 
-        return self._base.comps.environment_by_pattern(environment_name)
+        return environment_name
 
     def resolve_environment(self, environment_name):
         """Translate the given environment name to a group ID.
@@ -260,12 +274,9 @@ class DNFManager:
         :param environment_name: an identifier of an environment
         :return: a string with the environment ID or None
         """
-        env = self._get_environment(environment_name)
-
-        if not env:
-            return None
+        log.info("%s %s, environment %s" % (self.__class__.__name__, inspect.stack()[0][3], environment_name))
 
-        return env.id
+        return environment_name
 
     def get_environment_data(self, environment_name) -> CompsEnvironmentData:
         """Get the data of the specified environment.
@@ -287,24 +298,11 @@ class DNFManager:
         :param env: a DNF representation of the environment
         :return: an instance of CompsEnvironmentData
         """
+        task = self.tasks.get(env)
         data = CompsEnvironmentData()
-        data.id = env.id or ""
-        data.name = env.ui_name or ""
-        data.description = env.ui_description or ""
-
-        optional = {i.name for i in env.option_ids}
-        default = {i.name for i in env.option_ids if i.default}
-
-        for grp in self._base.comps.groups:
-
-            if grp.id in optional:
-                data.optional_groups.append(grp.id)
-
-            if grp.visible:
-                data.visible_groups.append(grp.id)
-
-            if grp.id in default:
-                data.default_groups.append(grp.id)
+        data.id = env or ""
+        data.name = task[0] or ""
+        data.description = task[1] or ""
 
         return data
 
@@ -314,7 +312,8 @@ class DNFManager:
 
         :return: a list of IDs
         """
-        return [g.id for g in self._base.comps.groups]
+        log.info("%s %s: %s" % (self.__class__.__name__, inspect.stack()[0][3], self.tasks.keys()))
+        return []
 
     def _get_group(self, group_name):
         """Translate the given group name into a DNF object.
@@ -359,9 +358,8 @@ class DNFManager:
         :return: an instance of CompsGroupData
         """
         data = CompsGroupData()
-        data.id = grp.id or ""
-        data.name = grp.ui_name or ""
-        data.description = grp.ui_description or ""
+        log.info("%s %s, %s" % (self.__class__.__name__, inspect.stack()[0][3], grp))
+        (data.id, data.name, data.description) = self.tasks[grp]
         return data
 
     def configure_proxy(self, url):
@@ -652,6 +650,11 @@ class DNFManager:
             args=(self._base, display)
         )
 
+        # create symlink /lib for distro feature usrmerge before transaction
+        # start since command depmod will create /lib as a directory
+        if not os.path.lexists(os.path.join(conf.target.system_root, 'lib')):
+            os.symlink("usr/lib", os.path.join(conf.target.system_root, "lib"))
+
         # Start the transaction.
         log.debug("Starting the transaction process...")
         process.start()
@@ -662,6 +665,15 @@ class DNFManager:
 
             # Wait for the transaction to end.
             process.join()
+
+            for pkg in self.failed_scriptlets_pkgnames.keys():
+                self._save_rpmpostinst(pkg)
+
+            # create symlinks for distro feature usrmerge
+            for d in ('bin', 'sbin', 'lib', 'lib32', 'lib64'):
+                if (not os.path.lexists(os.path.join(conf.target.system_root, d))
+                    and os.path.exists(os.path.join(conf.target.system_root, 'usr', d))):
+                    os.symlink(os.path.join('usr', d), os.path.join(conf.target.system_root, d))
         finally:
             # Kill the transaction after the timeout.
             process.join(timeout)
@@ -1023,3 +1035,116 @@ class DNFManager:
                 continue
 
         return ""
+
+    def environment_description(self, environment_id):
+        log.info("%s %s, %s" % (self.__class__.__name__, inspect.stack()[0][3], environment_id))
+
+        if environment_id not in self.tasks:
+            raise NoSuchGroup(environment_id)
+
+        if environment_id in self.tasks:
+            log.info("environmentDescription %s" % self.tasks)
+            (name, description, group) = self.tasks[environment_id]
+
+            return (name, description)
+        return (environment_id, environment_id)
+
+    def _oe_configure(self, dnf_conf):
+        # Refers oe-core's recipes of dnf and rpm
+        os.environ['RPM_NO_CHROOT_FOR_SCRIPTS'] = "1"
+
+        # Refers oe-core's meta/lib/oe/package_manager.py
+        # def _prepare_pkg_transaction(self) in class RpmPM(PackageManager)
+        target_rootfs = conf.target.system_root
+        os.environ['D'] = target_rootfs
+        os.environ['OFFLINE_ROOT'] = target_rootfs
+        os.environ['IPKG_OFFLINE_ROOT'] = target_rootfs
+        os.environ['OPKG_OFFLINE_ROOT'] = target_rootfs
+
+        # Refer oe-core's meta/lib/oe/package_manager.py
+        # def _configure_dnf(self) in class RpmPM(PackageManager)
+        # directly read its result from installer system
+        dnf_conf.substitutions['arch'] = open("/etc/dnf/vars/arch", "r").readline()
+
+        dnf_conf.errorlevel = dnf.const.VERBOSE_LEVEL
+        dnf_conf.debuglevel = dnf.const.VERBOSE_LEVEL
+
+    # OE specific
+    def read_buildstamp(self):
+        self.image, self.tasks = self._read_buildstamp()
+
+    def _read_buildstamp(self):
+        image = {}
+        tasks = {}
+
+        config = configparser.ConfigParser()
+        config.read(["/tmp/product/.buildstamp", "/.buildstamp", os.environ.get("PRODBUILDPATH", "")])
+
+        image_list = config.get("Rootfs", "LIST", fallback='').split()
+        for image_name in image_list:
+            _id_ = image_name.split('-')[0]
+            image_summary = config.get(image_name, "SUMMARY")
+            image_description = config.get(image_name, "DESCRIPTION")
+            package_install = config.get(image_name, "PACKAGE_INSTALL")
+            package_install_attemptonly = config.get(image_name,
+                                                         "PACKAGE_INSTALL_ATTEMPTONLY")
+            image_linguas = config.get(image_name, "IMAGE_LINGUAS")
+            image[image_name] = (image_summary,
+                                 image_description,
+                                 package_install,
+                                 package_install_attemptonly,
+                                 image_linguas)
+
+            short_image = image_name.replace("%s-image-" % _id_, "")
+
+            taskid = short_image
+            name = image_name
+            description = "%s" % image_summary
+            group = ""
+            tasks[taskid] = (name, description, group)
+
+            taskid = "%s-dev" % short_image
+            name = "%s dev-pkgs staticdev-pkgs" % image_name
+            description = "%s with development files" % image_summary
+            group = "dev-pkgs staticdev-pkgs"
+            tasks[taskid] = (name, description, group)
+
+            taskid = "%s-dbg" % short_image
+            name = "%s dbg-pkgs" % image_name
+            description = "%s with debug symbols" % image_summary
+            group = "dbg-pkgs"
+            tasks[taskid] = (name, description, group)
+
+            taskid = "%s-dev-dbg" % short_image
+            name = "%s dev-pkgs staticdev-pkgs dbg-pkgs" % image_name
+            description = "%s with development files and debug symbols" % image_summary
+            group = "dev-pkgs staticdev-pkgs dbg-pkgs"
+            tasks[taskid] = (name, description, group)
+
+        return image, tasks
+
+    # Refer upstream oe-core meta/lib/oe/package_manager.py
+    def _save_rpmpostinst(self, pkg):
+        def _script_num_prefix(path):
+            files = os.listdir(path)
+            numbers = set()
+            numbers.add(99)
+            for f in files:
+                numbers.add(int(f.split("-")[0]))
+            return max(numbers) + 1
+
+        log.info("Saving postinstall script of %s" % (pkg))
+
+        target_rootfs = conf.target.system_root
+        args = ["-q", "--root=%s" % target_rootfs, "--queryformat", "%{postin}", pkg]
+
+        output = execWithCapture('rpm', args, log_output=False)
+        log.info(output)
+
+        target_path = "%s/etc/rpm-postinsts" % target_rootfs
+        execWithRedirect("mkdir", ["-p", target_path])
+        num = _script_num_prefix(target_path)
+        saved_script_name = os.path.join(target_path, "%d-%s" % (num, pkg))
+        with open(saved_script_name, 'w') as f:
+            f.write(output)
+        os.chmod(saved_script_name, 0o755)
diff --git a/pyanaconda/modules/payloads/payload/dnf/initialization.py b/pyanaconda/modules/payloads/payload/dnf/initialization.py
index b983cf2506..5b10c68c67 100644
--- a/pyanaconda/modules/payloads/payload/dnf/initialization.py
+++ b/pyanaconda/modules/payloads/payload/dnf/initialization.py
@@ -212,6 +212,10 @@ class SetUpDNFSourcesTask(SetUpSourcesTask):
         dnf_manager.configure_substitution(self._release_version)
         dnf_manager.dump_configuration()
         dnf_manager.read_system_repositories()
+
+        # OE specific
+        dnf_manager.read_buildstamp()
+
         return dnf_manager
 
     def _load_source(self, dnf_manager, source, repository=None):
diff --git a/pyanaconda/modules/payloads/payload/dnf/installation.py b/pyanaconda/modules/payloads/payload/dnf/installation.py
index 8aefb95fe5..ee8df210e4 100644
--- a/pyanaconda/modules/payloads/payload/dnf/installation.py
+++ b/pyanaconda/modules/payloads/payload/dnf/installation.py
@@ -43,6 +43,7 @@ from pyanaconda.modules.payloads.payload.dnf.requirements import (
     collect_language_requirements,
     collect_platform_requirements,
     collect_remote_requirements,
+    collect_oe_requirements,
 )
 from pyanaconda.modules.payloads.payload.dnf.utils import (
     get_kernel_version_list,
@@ -161,7 +162,8 @@ class ResolvePackagesTask(CheckPackagesSelectionTask):
             + collect_language_requirements(self._dnf_manager) \
             + collect_platform_requirements(self._dnf_manager) \
             + collect_dnf_requirements(self._dnf_manager, self._configuration) \
-            + collect_driver_disk_requirements()
+            + collect_driver_disk_requirements() \
+            + collect_oe_requirements()
 
     def _collect_required_specs(self):
         """Collect specs for the required software."""
@@ -277,6 +279,12 @@ class InstallPackagesTask(Task):
         self._dnf_manager.install_packages(self.report_progress)
         return get_kernel_version_list()
 
+    def report_progress(self, message, step_number=None, step_size=None):
+        super().report_progress(message, step_number, step_size)
+
+        if message.startswith("Error in POSTIN scriptlet in rpm package"):
+            self._dnf_manager.failed_scriptlets_pkgnames[message.split()[-1]] = True
+
 
 class WriteRepositoriesTask(Task):
     """The installation task for writing repositories on the target system."""
@@ -457,3 +465,67 @@ class UpdateDNFConfigurationTask(Task):
         if rc != 0:
             log.warning("Failed to update the DNF configuration (%s).", rc)
             return
+
+
+class OEConfigureTask(Task):
+    """The OE configuration task."""
+
+    def __init__(self, sysroot, dnf_module):
+        """Create a new task.
+
+        :param str sysroot: a path to the system root
+        :param DNFManager dnf_manager: a DNF manager
+        """
+        super().__init__()
+        self._sysroot = sysroot
+        self._dnf_module = dnf_module
+
+    @property
+    def name(self):
+        return "Update DNF configuration for OE"
+
+    def run(self):
+        """Run the task."""
+        self._oe_configure_post()
+
+        # For the generation of anaconda-ks.cfg
+        pkg_data = self._dnf_module.packages_selection
+        for pkg in self._list_installed():
+            if pkg not in pkg_data.packages:
+                pkg_data.packages.append(pkg)
+        self._dnf_module.set_packages_selection(pkg_data)
+
+        util.execWithRedirect("depmod",
+                              ["-a"],
+                              root=conf.target.system_root)
+
+    def _oe_configure_post(self):
+        # Refers oe-core's meta/lib/oe/package_manager.py
+        # def _configure_rpm(self) and def _configure_dnf(self)
+        # in class RpmPM(PackageManager)
+        # directly copy its result from installer system
+        target_sysconf = "%s/etc" % conf.target.system_root
+        util.execWithRedirect("cp", ["-fr", "/etc/rpmrc", target_sysconf])
+        util.execWithRedirect("cp", ["-rf", "/etc/dnf", target_sysconf])
+        util.execWithRedirect("cp", ["-rf", "/etc/rpm", target_sysconf])
+
+    def _list_installed(self):
+        output = util.execWithCapture(
+                         'dnf',
+                         [
+                           'repoquery',
+                           '--installed',
+                           '--installroot=' + self._sysroot,
+                           '--queryformat=Package:%{name}'
+                         ],
+                         log_output=False
+                 )
+        output = output.rstrip()
+        lines = output.splitlines()
+        installed = []
+        for line in lines:
+            if not line.startswith("Package:"):
+                continue
+            installed.append(line.replace("Package:", ""))
+
+        return installed
diff --git a/pyanaconda/modules/payloads/payload/dnf/requirements.py b/pyanaconda/modules/payloads/payload/dnf/requirements.py
index 8f84292103..b3b3675e1a 100644
--- a/pyanaconda/modules/payloads/payload/dnf/requirements.py
+++ b/pyanaconda/modules/payloads/payload/dnf/requirements.py
@@ -154,6 +154,27 @@ def collect_driver_disk_requirements(path="/run/install/dd_packages"):
     return requirements
 
 
+def collect_oe_requirements():
+    """Collect the requirements for oe
+    """
+    requirements = []
+
+    # base packages
+    for pkg in ('base-files', 'base-passwd', 'shadow'):
+        requirements.append(Requirement.for_package(
+            pkg,
+            reason='base packages'
+        ))
+    # Support grub-mkconfig
+    for pkg in ('sed', 'coreutils'):
+        requirements.append(Requirement.for_package(
+            pkg,
+            reason='support grub-mkconfig'
+        ))
+
+    return requirements
+
+
 def apply_requirements(requirements, include_list, exclude_list):
     """Apply the provided requirements.
 
diff --git a/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py b/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py
index f0d77e1448..0308b81f32 100644
--- a/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py
+++ b/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py
@@ -50,6 +50,9 @@ def process_transaction_progress(queue, callback):
         elif token == 'quit':
             log.info(msg)
             break  # Installation finished successfully
+        elif token == 'warn':
+            if msg.startswith("Error in POSTIN scriptlet in rpm package"):
+                callback(msg)
         elif token == 'error':
             log.error(msg)
             raise PayloadInstallationError("An error occurred during the transaction: " + msg)
@@ -123,7 +126,11 @@ class TransactionProgress(dnf.callback.TransactionProgress):
 
         :param message: a string that describes the error
         """
-        self._queue.put(('error', message))
+        token = "error"
+        if message.startswith("Error in POSTIN scriptlet in rpm package"):
+            token = "warn"
+        self._queue.put((token, message))
+        log.info("error: %s" % message)
 
     def quit(self, message):
         """Report the end of the transaction and close the queue.
diff --git a/pyanaconda/modules/payloads/payload/dnf/utils.py b/pyanaconda/modules/payloads/payload/dnf/utils.py
index fe47d4ff67..830cd5c45c 100644
--- a/pyanaconda/modules/payloads/payload/dnf/utils.py
+++ b/pyanaconda/modules/payloads/payload/dnf/utils.py
@@ -112,18 +112,8 @@ def get_installation_specs(data: PackagesSelectionData, default_environment=None
     exclude_list = []
 
     # Handle the environment.
-    if data.default_environment_enabled and default_environment:
-        log.info("Selecting default environment '%s'.", default_environment)
-        include_list.append("@{}".format(default_environment))
-    elif data.environment:
-        include_list.append("@{}".format(data.environment))
 
     # Handle the core group.
-    if not data.core_group_enabled:
-        log.info("Skipping @core group; system may not be complete.")
-        exclude_list.append("@core")
-    else:
-        include_list.append("@core")
 
     # Handle groups.
     for group_name in data.excluded_groups:
diff --git a/pyanaconda/modules/payloads/payload/dnf/validation.py b/pyanaconda/modules/payloads/payload/dnf/validation.py
index d061f509de..339a563091 100644
--- a/pyanaconda/modules/payloads/payload/dnf/validation.py
+++ b/pyanaconda/modules/payloads/payload/dnf/validation.py
@@ -17,6 +17,7 @@
 #
 from contextlib import contextmanager
 
+from pyanaconda.core import util
 from pyanaconda.anaconda_loggers import get_module_logger
 from pyanaconda.core.i18n import _
 from pyanaconda.modules.common.structures.packages import PackagesSelectionData
@@ -117,6 +118,17 @@ class CheckPackagesSelectionTask(ValidationTask):
             self._selection, default_environment
         )
 
+        env = None
+        if self._selection.environment:
+            env = self._selection.environment
+        elif default_environment:
+            env = default_environment
+        log.info("selected environment: %s", env)
+
+        for pkg_name in self._apply_yocto_environment(env):
+            log.info("selected yocto package: '%s'", pkg_name)
+            include_list.append(pkg_name)
+
         self._include_list.extend(include_list)
         self._exclude_list.extend(exclude_list)
 
@@ -158,3 +170,74 @@ class CheckPackagesSelectionTask(ValidationTask):
             report.error_messages.append(str(e))
         except InvalidSelectionError as e:
             report.error_messages.append(str(e))
+
+    def _apply_yocto_environment(self, env):
+        if not env:
+            return []
+
+        (image_name, description) = self._dnf_manager.environment_description(env)
+        log.info("image_name %s, des %s" % (image_name, description))
+        image_id = image_name.split()[0]
+        (image_summary, image_description, package_install,
+            package_install_attemptonly, image_linguas) = self._dnf_manager.image[image_id]
+        log.info("package_install %s, attempt %s, linguas %s" %
+                 (package_install, package_install_attemptonly, image_linguas))
+
+        # Language locale
+        self._select_yocto_linguas(image_linguas)
+
+        # -dbg/-doc/-dev
+        for group in self._dnf_manager.tasks[env][2].split():
+            self._select_yocto_group(group)
+
+        # Installing packages
+        return package_install.split()
+
+    def _select_yocto_group(self, group_id):
+        log.info("_select_yocto_group %s" % group_id)
+        if not group_id:
+            return []
+
+        glob = self._yocto_complementary_glob(group_id)
+
+        # get all available languages in repos
+        available_packs = self._dnf_manager._base.sack.query().available().filter(
+                            name__glob="*-%s" % glob)
+        allpkgnames = [p.name for p in available_packs]
+        log.info("allpkgnames %d" % len(allpkgnames))
+        for pkgname in allpkgnames:
+            self._try_install_package(pkgname)
+
+    def _select_yocto_linguas(self, image_linguas):
+        lc_globs = []
+        for ling in image_linguas.split():
+            if ling not in lc_globs:
+                lc_globs.append(ling)
+
+            baselang = ling.split('-')[0]
+            if ling != baselang and baselang not in lc_globs:
+                lc_globs.append(baselang)
+
+        log.info("lc_globs: %s" % lc_globs)
+        for lc_glob in lc_globs:
+            self._select_yocto_group("locale-%s" % lc_glob)
+
+    def _try_install_package(self, pkg_name):
+        try:
+            return self._dnf_manager._base.install(pkg_name, strict=False)
+        except dnf.exceptions.MarkingError:
+            msg = "Payload error - No such package: %s" % pkg_name
+            raise PayloadError(msg)
+
+    def _yocto_complementary_glob(self, group):
+        complementary_glob = {}
+        complementary_glob['dev-pkgs'] = 'dev'
+        complementary_glob['staticdev-pkgs'] = 'staticdev'
+        complementary_glob['doc-pkgs'] = 'doc'
+        complementary_glob['dbg-pkgs'] = 'dbg'
+        complementary_glob['ptest-pkgs'] = 'ptest'
+
+        if group in complementary_glob:
+            return complementary_glob[group]
+
+        return group
diff --git a/pyanaconda/payload/dnf/payload.py b/pyanaconda/payload/dnf/payload.py
index 096bf86b0a..ba90efea2f 100644
--- a/pyanaconda/payload/dnf/payload.py
+++ b/pyanaconda/payload/dnf/payload.py
@@ -118,8 +118,8 @@ class DNFPayload(MigratedDBusPayload):
             set_source(self.proxy, source_proxy)
 
         else:
-            log.debug("Use the DNF source from the Anaconda configuration file.")
-            source_proxy = create_source(conf.payload.default_source)
+            log.debug("Use the DNF source from install disk.")
+            source_proxy = create_source(SOURCE_TYPE_CDROM)
             set_source(self.proxy, source_proxy)
 
     @staticmethod
-- 
2.34.1
